### ISSUE TITLE
A (light) refactor inspired by modern compiler theory

### ISSUE THREAD

**Issue by schneiderfelipe** [cclib/cclib#1124]
> TL;DR: Let's make cclib closer to a compiler and improve [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) in the process!
> This is *not* an Abril's fool joke.

In my view, some features are hard to implement with the current architecture. This is mainly due to a sort of information duplication shared by all parsers: they all have to know about the final data structure, which is [part of the public API](https://cclib.github.io/data.html). Why is this information duplication? Because if we ever want to change `ccData`, we have to change every single parser.

For instance, #89 or #988 would require updating all parsers just to change units in `ccData`. And it seems reasonable to me that parsers should have no responsibility for units in `ccData` (other than the ones outputted by each computational chemistry package). To me, this is a case of double duty.

Even if we don't change its structure, manipulating `ccData` before returning it has to be done at the parser level. This makes it harder to implement things like #50 or #789 ([see also this comment](https://github.com/cclib/cclib/issues/789#issuecomment-1002230073) by @awvwgk), since it would require parsing a second file, and this would return another `ccData`. It would have to be merged within the requesting parser itself. So parsers would have to both consume and produce `ccData`, which seems to be a double duty as well.

An example of that can be found in the excellent work of @shivupa in #1076. Ideally, we would like to reuse parsers (which would help with #808 and #1012, for instance), but as they accumulate more and more responsibilities (and complexity, and *statefulness*), this becomes harder and harder.

To solve those issues, we could use some ideas from modern compiler theory.

## What I'm proposing

> This is related to #335.

I propose creating the `ccData` object *outside* parsers. This is similar to [how most compilers work](https://en.wikipedia.org/wiki/Compiler#Front_end). Our parsers become [tokenizers](https://en.wikipedia.org/wiki/Lexical_analysis), i.e., [iterators emitting *tokens*](https://stackoverflow.com/a/15895283/4039050) (a.k.a. *events*) as the log file is read. A [package-agnostic object](https://en.wikipedia.org/wiki/Parsing) then consumes this token stream and produces a `ccData` object. The main point is that [iterators](https://docs.python.org/3/glossary.html#term-iterator) are easy to compose, so multi-job/multi-file can be done in a general way.

What would change? Code like https://github.com/cclib/cclib/blob/5798fdcdd91813cbdfb99cb36a2e48c16be9a155/cclib/parser/orcaparser.py#L736-L738 would change to

```python
    yield {"set_attribute": "natom", "value": len(atomnos)}
    yield {"set_attribute": "atomnos", "value": atomnos}
    yield {"append_attribute": "atomcoords", "value": atomcoords}
```

(Or something equivalent with a dedicated class for tokens.) As long as the token stream consumer produces a valid `ccData`, nothing would break. But the composability greatly increases. For instance, we could transform https://github.com/cclib/cclib/blob/693bd40f5b9956608884531d7e70bb60ca31027f/cclib/io/ccio.py#L79-L100 into a special tokenizer that detects which log file we have, which could trigger a tokenizer switch.

## How does this solve anything?

Examples:
- Dedicated token stream consumers (i.e., *the actual* parser) could filter tokens and only produce user-requested data or format (#335).
- Token streams can be manipulated before returning. For instance, units could be requested by the caller (#998).
- Nested and multiple-job log files become easier by chaining/interspersing token streams from different sources (#1076).

**Comment by awvwgk** [comment]
Sorry for hopping in here, this somehow landed in my inbox.

I wouldn't call such a change a compiler, you are more likely to separate your existing parser in separate stages, *i.e.* a scanner and an actual parser. Instead of having a parser for every file format you seem to plan on shifting to scanner for every format and one central parser to make sense of everything.

Diving into compiler / interpreter theory is fun, I can recommend [crafting interpreters](https://craftinginterpreters.com), but it can become tempting to apply language construction concepts to every problem ;).

**Comment by schneiderfelipe** [comment]
> I wouldn't call such a change a compiler, you are more likely to separate your existing parser in separate stages, _i.e._ a scanner and an actual parser. Instead of having a parser for every file format you seem to plan on shifting to scanner for every format and one central parser to make sense of everything.

Yes, this is exactly what I'm proposing. I think this might make future changes a lot easier.

**Comment by langner** [comment]
I like the general thrust here, in the sense of separating concerns between parsing and building the data object, since it allows some more flexibility. I'll punt to 2.x, and we should consider what part of this makes the most sense in a future where all attributes are classes (the main change going into 2.x).
